README
= Events Calendar

A calendar of events for Radiant CMS.

Note: This version of the events_calendar extension is compatible with Radiant 0.8.
If you are using Radiant 0.7.1, please pull from the radiant-0.7.1 branch.

WARNING: The syntax of the `<r:events>` tag changed between version 0.3 and 0.4.

A bare `<r:events>` tag no longer returns the events for the current day (use
`<r:events for="today">` to achieve the same results) and instead returns all
events.

The `year`, `month`, and `day` attributes are no longer used; use the `for`
attribute instead. Instead of

  <r:events year="2009" month="12" day="6">

use

  <r:events for="2009-12-06">

== Installation

Install the events_calendar extension:

  git clone git://github.com/davec/radiant-events-calendar-extension.git vendor/extensions/events_calendar

Now, run the migration and copy the public assets:

  rake radiant:extensions:events_calendar:update
  rake radiant:extensions:events_calendar:migrate

To use the calendar, you must include the provided JavaScript and CSS files in
your layout. The included JavaScript relies on the Prototype and LowPro
libraries, so they must also be loaded prior to loading the events_calendar.js
file. Add the following to your layout header:

  <link rel="stylesheet" type="text/css" href="/stylesheets/events_calendar.css" />
  <script type="text/javascript" src="/javascripts/prototype.js"></script>
  <script type="text/javascript" src="/javascripts/lowpro.js"></script>
  <script type="text/javascript" src="/javascripts/events_calendar.js"></script>

An alternate version of the JavaScript file, named events_calendar_jquery.js,
is provided for use with jQuery (the minimum jQuery version is 1.4.2 and is
not included).

Events can be created and manged from the Events tab in the admin UI. A
calendar and listings of events can be displayed in your templates using the
supplied Radius tags.

== Configuration

This extension uses a custom page type called Events that is used to show all
events for a specific day via the /events/year/month/day path. (This replaces
the events show template that was previously used.) A default events page is
installed unless a page with a slug named `events` already exists.

When used in an Events page, the <r:events> tag does not take any attributes.

== Example

To display a calendar of the current month, use

  <r:calendar />

The month and year tags can be added to display a calendar for a specific month.

  <r:calendar year='2008' month='8' />

You can iterate over all events like so:

  <r:events>
    <dl>
      <r:each>
        <dt><r:event:name /></dt>
        <dd><r:event:time /></dd>
        <dd><r:event:location /></dd>
        <dd><r:event:description /></dd>
      </r:each>
    </dl>
  </r:events>

For today's events:

  <r:events for="today">
    <r:each>
      ...
    </r:each>
  </r:events>

For some other day's events:

  <r:events for="2009-03-14">
    <r:each>
      ...
    </r:each>
  </r:events>

Other options that you can put in the events tag (similar to children:each):
* by (default: start_date)
* order (default: asc)
* limit
* offset
* category

Special dates in the `for` option:
* today
* tomorrow
* yesterday

Relative date periods can also be used in the `for` option:
* next 1 week
* next 3 days
* previous 2 months
* any period matching the regexp `/(next|previous) \d+ (day|week|month|year)s?/`

When specifying a relative date period, the `inclusive` attribute can also be
specified to control whether the current day is included in the range. The
default setting `inclusive="true"` included the current day, but if `inclusive`
is set to any other value the current day is excluded from the returned value.

== Dependencies

This extension depends on, and includes, the following plugins:

* auto_complete <http://github.com/rails/auto_complete>
* calendar_date_select <http://github.com/timcharper/calendar_date_select>
* calendar_helper <http://github.com/topfunky/calendar_helper>
* simple_time_select <http://github.com/tamoyal/simple_time_select>

== Contributors

David Cato (davec)
Jeremy Stephens (viking)


CHANGELOG
== v0.9.3

* Improved test coverage [David Cato]

* Assign attrs before use in truncate_html [David Cato]

* Re-order response targets [David Cato]

  Check for html before js

* Warn when end_time < start_time [David Cato]

* Add more admin controller tests [David Cato]

* Fix end time adjustment for events on last day of month [David Cato]

* Allow events to end after midnight [David Cato]

  If end_time < start_time, advance end_time by 1 day

* Add padding to calendar to avoid accordion effects [David Cato]

* Remove empty table rows from the generated calendar [David Cato]

* Remove unused rjs param in make_calendar [David Cato]

* Revert "Fix for IE" [David Cato]

  Doesn't seem to actually be needed; problem was apparently elsewhere.
  Besides, the intended fix was accidently left out.

== v0.9.2

* Update event form to work with textile toolbar [David Cato]

* temporary hack to allow for textile toolbar [Jim Gay]

* Fix for IE [David Cato]

* Make the public-facing calendar work with jQuery [David Cato]

== v0.9.1

* Force setting of page filter when creating event [David Cato]

  Need to have a known page filter defined for events so that the
  rendering is consistent, regardless of the page filter setting in the
  containing project.

* DRY up calendar display with lowpro behaviors [David Cato]

* Remove unrecognized format attr from <r:description> [David Cato]

* Add <r:event:time:start> and <r:event:time:end> tags [David Cato]

* Fix inline code strings in tag description text [David Cato]

* Add a custom Events page type [David Cato]

  This replaces the events controller and associated show view and drops
  the dependency on shared_layouts. Changes to the events page can now be
  made easily without requiring a re-deploy.

  Added additional tags (<r:event:if_time>, <r:event:if_location>,
  <r:event:if_description>, <r:event:if_category>, and their unless
  counterparts) and changed <r:event:description> to return the
  HTML-formatted description instead of the raw description.

* Add CHANGELOG [David Cato]

* Adjust default event ordering [David Cato]

* Fix typo [David Cato]

* Move sync task to the proper level [David Cato]

* Add missing load of words in sync task [David Cato]

* Tags get i18n treatment [David Cato]

* Update tag descriptions [David Cato]

* Minor event form tweaks [David Cato]

* Update filter reference display [David Cato]

* Make auto-completion results visible [David Cato]

== v0.9

* Add i18n (admin UI only)

== v0.8.3

* Improved test coverage [David Cato]

* Assign attrs before use in truncate_html [David Cato]

* Re-order response targets [David Cato]

  Check for html before js

* Warn when end_time < start_time [David Cato]

* Add more admin controller tests [David Cato]

* Fix end time adjustment for events on last day of month [David Cato]

* Allow events to end after midnight [David Cato]

  If end_time < start_time, advance end_time by 1 day

* Add padding to calendar to avoid accordion effects [David Cato]

* Remove empty table rows from the generated calendar [David Cato]

* Remove unused rjs param in make_calendar [David Cato]

* Revert "Fix for IE" [David Cato]

  Doesn't seem to actually be needed; problem was apparently elsewhere.
  Besides, the intended fix was accidently left out.

== v0.8.2

* Update event form to work with textile toolbar [David Cato]

* temporary hack to allow for textile toolbar [Jim Gay]

* Fix for IE [David Cato]

* Make the public-facing calendar work with jQuery [David Cato]

== v0.8.1

* DRY up calendar display with lowpro behaviors [David Cato]

* Remove unrecognized format attr from <r:description> [David Cato]

* Add <r:event:time:start> and <r:event:time:end> tags [David Cato]

* Fix inline code strings in tag description text [David Cato]

* Add a custom Events page type [David Cato]

  This replaces the events controller and associated show view and drops
  the dependency on shared_layouts. Changes to the events page can now be
  made easily without requiring a re-deploy.

  Added additional tags (<r:event:if_time>, <r:event:if_location>,
  <r:event:if_description>, <r:event:if_category>, and their unless
  counterparts) and changed <r:event:description> to return the
  HTML-formatted description instead of the raw description.

* Add CHANGELOG [David Cato]

* Adjust default event ordering [David Cato]

* controls.js -> admin/controls.js [David Cato]

* Convert build_select options param to String [David Cato]

  Syntax change in Rails 2.3.8 [David Cato]

* Update admin_reference_path call [David Cato]

  The associated route changed in Radiant 0.9 [David Cato]

* Update tabs for Radiant 0.9 [David Cato]

* Move route defs to config/routes.rb [David Cato]

== v0.8

* Version bump

== v0.5

* Add event copy functionality [David Cato]
    
  Create a new event based on an existing event. This is a quick-and-dirty
  way of providing multi-day events; specify the details for the first day
  and make copies for subsequent days, changing the date for each copy.

* Clean up tests  [David Cato]

* Fix calendar navigation when JavaScript is disabled  [David Cato]

* Fix remove error. Closes #2  [David Cato]

* Add relative time periods for event date spec  [David Cato]

  The `for' attribute now accepts relative time periods of the format
  /(next|previous) \d+ (day|month|week|year)s?/
    
  So the following can now be specified:
    
  <r:events for="next 2 weeks" /> or <r:events for="previous 1 month" />
    
  By default, today is included, but if the `inclusive' attribute is passed
  the (string literal) value of anything other than "true" then today is
  not included in the result.

* Set start_time to match date [David Cato]

  Two events used Date.today.year for the date but hard-coded year 2009
  for the start_time.

* Clean up JS code [David Cato]

  Declare vars and don't initialize the last filter name to the default
  filter since doing so prevents the filter help from being displayed

app/models/events_calendar_tags.rb
module EventsCalendarTags
  include Radiant::Taggable
  include CalendarsHelper

  class TagError < StandardError ; end

  desc %{
    Gives access to all events, sorted by start_time by default.

    The <code>for</code> attribute can be any of the following:
    * "today"
    * "tomorrow"
    * "yesterday"
    * a date in YYYY-MM-DD format (e.g., 2009-03-14)
    * a relative number of days, weeks, months, or years from the current day (e.g., "next 2 weeks", "previous 7 days")

    The <code>inclusive</code> attribute applies to relative <code>for</code> values. If set to <code>true</code> (the default) then <code>today</code> is included; otherwise <code>today</code> is excluded.

    Events can be sorted (with the <code>by</code> attribute) by name, location, date,
    start_time, or end_time; the order of sorting is determined by the <code>order</code>
    attribute (default sorting is in ascending order by event start time).

    *Usage:*

    <pre><code><r:events [for="date" [inclusive="true|false"]] [by="attribute"] [order="asc|desc"] [limit="number"] [offset="number"]/></code></pre>
  }
  tag 'events' do |tag|
    tag.locals.events = Event.all(events_find_options(tag))
    tag.expand
  end

  desc %{
    Loops through events.
  }
  tag 'events:each' do |tag|
    tag.locals.events.collect do |event|
      tag.locals.event = event
      tag.expand
    end
  end

  desc %{
    Creates the context for a single event.
  }
  tag 'events:each:event' do |tag|
    tag.expand
  end

  desc %{
    Renders the name for the current event.
  }
  tag 'event:name' do |tag|
    tag.locals.event.name
  end

  desc %{
    Renders the date for the current event.

    An optional date format string, as used for <code>Date#strftime</code>, may be
    specified with the <code>format</code> attribute (the default <code>%Y-%m-%d</code>).

    *Usage:*
    <pre><code><r:event:date /></code></pre>
    <pre><code><r:event:date format='%d %b %Y' /></code></pre>
  }
  tag 'event:date' do |tag|
    format = tag.attr['format'] || '%Y-%m-%d'
    tag.locals.event.date.strftime(format)
  end

  desc %{
    Renders the time for the current event.

    An optional time format string, as used for <code>Date#strftime</code>, may be
    specified with the <code>format</code> attribute (the default is <code>%H:%M</code>).
    
    For events with both start and end time values, the <code>connector</code> attribute
    sets the text that separates the start and end time strings (the default
    is a single hyphen).

    *Usage:*
    <pre><code><r:event:time /></code></pre>
    <pre><code><r:event:time format='%I:%M %p' connector='to' /></code></pre>
  }
  tag 'event:time' do |tag|
    return tag.expand if tag.double?

    event = tag.locals.event
    return '' unless event.start_time

    format = tag.attr['format'] || '%H:%M'
    [ event.start_time.strftime(format),
      event.end_time ? event.end_time.strftime(format) : nil ].compact.join(" #{tag.attr['connector'] || '-'} ")
  end

  desc %{
    Render inner content if the current event has a defined start time.

    *Usage:*
    <pre><code><r:event:if_time>...</r:event:if_time></code></pre>
  }
  tag 'event:if_time' do |tag|
    tag.expand if tag.locals.event.start_time
  end

  desc %{
    Render inner content if the current event does not have a defined start time.

    *Usage:*
    <pre><code><r:event:unless_time>...</r:event:unless_time></code></pre>
  }
  tag 'event:unless_time' do |tag|
    tag.expand unless tag.locals.event.start_time
  end

  desc %{
    Renders the start time for the current event.
    The default time format is <code>%H:%M</code> but can be changed by setting a different format with the <code>format</code> attribute.

    *Usage:*
    <pre><code><r:event:time:start /></code></pre>
    <pre><code><r:event:time:start format='%I:%M %p' /></code></pre>
  }
  tag 'event:time:start' do |tag|
    event = tag.locals.event
    return '' unless event.start_time

    format = tag.attr['format'] || '%H:%M'
    event.start_time.strftime(format)
  end

  desc %{
    Renders the end time for the current event.
    The default time format is <code>%H:%M</code> but can be changed by setting a different format with the <code>format</code> attribute.

    *Usage:*
    <pre><code><r:event:time:end /></code></pre>
    <pre><code><r:event:time:end format='%I:%M %p' /></code></pre>
  }
  tag 'event:time:end' do |tag|
    event = tag.locals.event
    return '' unless event.end_time

    format = tag.attr['format'] || '%H:%M'
    event.end_time.strftime(format)
  end

  desc %{
    Renders the location for the current event.
  }
  tag 'event:location' do |tag|
    tag.locals.event.location
  end

  desc %{
    Render inner content if the current event has a location.

    *Usage:*
    <pre><code><r:event:if_location>...</r:event:if_location></code></pre>
  }
  tag 'event:if_location' do |tag|
    tag.expand if tag.locals.event.location
  end

  desc %{
    Render inner content if the current event does not have a location.

    *Usage:*
    <pre><code><r:event:unless_location>...</r:event:unless_location></code></pre>
  }
  tag 'event:unless_location' do |tag|
    tag.expand unless tag.locals.event.location
  end

  desc %{
    Renders the HTML-formatted description for the current event.

    The returned HTML is sanitized for your protection using <code>HTML::WhiteListSanitizer#sanitize</code>.
    If sanitization is not desired, specify <code>sanitize="false"</code>.

    *Usage:*
    <pre><code><r:event:description /></code></pre>
    <pre><code><r:event:description sanitize="false" /></code></pre>
  }
  tag 'event:description' do |tag|
    event = tag.locals.event
    sanitize = tag.attr['sanitize'] || 'true'
    case sanitize
    when 'false'
      event.description_html
    when 'true'
      @sanitizer ||= HTML::WhiteListSanitizer.new
      @sanitizer.sanitize(event.description_html)
    else
      raise TagError, %{the `sanitize' attribute of the `description' tag must be either "true" or "false"}
    end
  end

  desc %{
    Render inner content if the current event has a description.

    *Usage:*
    <pre><code><r:event:if_description>...</r:event:if_description></code></pre>
  }
  tag 'event:if_description' do |tag|
    tag.expand if tag.locals.event.description
  end

  desc %{
    Render inner content if the current event does not have a description.

    *Usage:*
    <pre><code><r:event:unless_description>...</r:event:unless_description></code></pre>
  }
  tag 'event:unless_description' do |tag|
    tag.expand unless tag.locals.event.description
  end

  desc %{
    Renders the category for the current event.
  }
  tag 'event:category' do |tag|
    tag.locals.event.category
  end

  desc %{
    Render inner content if the current event has a category.

    *Usage:*
    <pre><code><r:event:if_category>...</r:event:if_category></code></pre>
  }
  tag 'event:if_category' do |tag|
    tag.expand if tag.locals.event.category
  end

  desc %{
    Render inner content if the current event does not have a category.

    *Usage:*
    <pre><code><r:event:unless_category>...</r:event:unless_category></code></pre>
  }
  tag 'event:unless_category' do |tag|
    tag.expand unless tag.locals.event.category
  end

  desc %{
    Creates a calendar for the given month.
    If no date is given, the current month is used.

    *Usage:*
    <pre><code><r:calendar /></code></pre>
    <pre><code><r:calendar year='2009' month='1' /></code></pre>
  }
  tag 'calendar' do |tag|
    options = [ tag.attr['year'], tag.attr['month'] ].compact
    raise TagError, "the calendar tag requires a month and year to be specified" unless options.empty? || options.length == 2

    # If the JavaScript is not enabled in the client, CalendarsController#show stores the calendar dates in the session
    stored_calendar_view = request.session[:calendar_view] || {}

    year  = tag.attr['year']  || stored_calendar_view[:year]  || Date.today.year
    month = tag.attr['month'] || stored_calendar_view[:month] || Date.today.month

    request.session.delete(:calendar_view)

    date = Date.civil(year.to_i, month.to_i)
    make_calendar(date)
  end

  private

    def events_find_options(tag)
      attr = tag.attr.symbolize_keys

      options = {}
      where_clauses = []
      where_values = []

      [:limit, :offset].each do |symbol|
        if number = attr[symbol]
          if number =~ /^\d{1,4}$/
            options[symbol] = number.to_i
          else
            raise TagError, %{the `#{symbol}' attribute of the `each' tag must be a positive number between 1 and 4 digits}
          end
        end
      end

      by = (attr[:by] || 'start_time').strip
      order = (attr[:order] || 'asc').strip
      order_string = ""
      if Event.column_names.include?(by)
        order_string << by
      else
        raise TagError, %{the `by' attribute of the `each' tag must be set to a valid field name}
      end
      if order =~ /^(asc|desc)$/i
        order_string << " #{$1.upcase}"
      else
        raise TagError, %{`order' attribute of `each' tag must be set to either "asc" or "desc"}
      end
      options[:order] = order_string

      if attr[:for]
        if attr[:for] =~ /\A(next|previous)\s+(\d+)\s+(day|week|month|year)s?\z/
          direction, count, interval = $1, $2.to_i, $3
          include_today = (attr[:inclusive] || "true") == "true"
          if direction == "next"
            start_date = include_today ? Date.today : Date.today + 1
            end_date = Date.today + count.send(interval)
          else
            start_date = Date.today - count.send(interval)
            end_date = include_today ? Date.today : Date.today - 1
          end

          where_clauses << "date >= ?" << "date <= ?"
          where_values << start_date << end_date
        else
          date = case attr[:for]
                 when "today"
                   Date.today
                 when "yesterday"
                   Date.today - 1
                 when "tomorrow"
                   Date.today + 1
                 else
                   parts = attr[:for].split("-")
                   raise TagError, "invalid date" unless parts.length == 3
                   Date.civil(*parts.collect(&:to_i))
                 end

          where_clauses << "date = ?"
          where_values  << date
        end
      end

      if attr[:category]
        where_clauses << "category = ?"
        where_values  << attr[:category]
      end

      if !where_clauses.empty?
        options[:conditions] = [where_clauses.join(" AND ")] + where_values
      end

      options
    end
end
